### PR TITLE
test: replace hashlib.ripemd160 with an own implementation

### LINKED ISSUE #23710: test: Openssl removed ripemd160
Openssl removed ripemd160, see https://github.com/openssl/openssl/issues/16994

Thus, our tests fail. See below for excerpt.

Not sure how to fix this, so leaving an issue here.

```
Running Unit Tests for Test Framework Modules
..........
----------------------------------------------------------------------
Ran 10 tests in 0.812s

OK
4/225 - feature_block.py failed, Duration: 0 s                                                                                                                                          

stdout:


stderr:
Traceback (most recent call last):
  File "/usr/lib/python3.9/hashlib.py", line 160, in __hash_new
    return _hashlib.new(name, data, **kwargs)
ValueError: [digital envelope routines] initialization error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/bitcoin-core/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/feature_block.py", line 54, in <module>
    from data import invalid_txs
  File "/bitcoin-core/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/data/invalid_txs.py", line 57, in <module>
    basic_p2sh = script_to_p2sh_script(CScript([OP_0]))
  File "/bitcoin-core/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/script_util.py", line 74, in script_to_p2sh_script
    return scripthash_to_p2sh_script(hash160(script))
  File "/bitcoin-core/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/script.py", line 35, in hash160
    return hashlib.new('ripemd160', sha256(s)).digest()
  File "/usr/lib/python3.9/hashlib.py", line 166, in __hash_new
    return __get_builtin_constructor(name)(data)
  File "/usr/lib/python3.9/hashlib.py", line 123, in __get_builtin_constructor
    raise ValueError('unsupported hash type ' + name)
ValueError: unsupported hash type ripemd160



TEST                                               | STATUS    | DURATION

wallet_backup.py --descriptors                     | ✓ Passed  | 109 s
wallet_hd.py --descriptors                         | ✓ Passed  | 92 s
wallet_hd.py --legacy-wallet                       | ✓ Passed  | 125 s
feature_block.py                                   | ✖ Failed  | 0 s

ALL                                                | ✖ Failed  | 326 s (accumulated) 
Runtime: 125 s

### PR DESCRIPTION
**Pull request by sipa**
Closes #23710.

### CODE REVIEW

**Comment by jamesob** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/23716/commits/5b559dc7ecf37ab1604b75ec8ffe8436377a5fb1, pending CI

Verified test vectors match my system's ripemd160 installation. Did not review the actual crypto.

File: test/functional/test_framework/ripemd160.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
+# Copyright (c) 2021 Pieter Wuille
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test-only pure Python RIPEMD160 implementation."""
```
Comment (luke-jr) [inline test/functional/test_framework/ripemd160.py:4]: What makes this inherently test-only?

Comment (sipa) [reply test/functional/test_framework/ripemd160.py:4]: It's not constant time.

Comment (luke-jr) [reply test/functional/test_framework/ripemd160.py:4]: IMO should be mentioned in the comment.

**Comment by MarcoFalke** [comment]
After merge there will be one other remaining use of `new()`. (`git grep 'hashlib.new'`). Since it is discouraged in the python docs, we should replace that with the named constructor.

Going to merge this, but I plan to review after merge.

File: test/functional/test_framework/ripemd160.py (hunk 1 above)
Comment (Sjors) [reply test/functional/test_framework/ripemd160.py:4]: That's indeed a good warning to add in a followup.

File: test/functional/test_framework/ripemd160.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+def ripemd160(data):
+    """Compute the RIPEMD-160 hash of data."""
+    # Initialize state.
+    state = (0x67452301, 0xefcdab89, 0x98badcfe, 0x10325476, 0xc3d2e1f0)
+    # Process full 64-byte blocks in the input.
+    for b in range(len(data) >> 6):
+        state = compress(*state, data[64*b:64*(b+1)])
+    # Construct final blocks (with padding and size).
+    pad = b"\x80" + b"\x00" * ((119 - len(data)) & 63)
```
Comment (MarcoFalke) [inline test/functional/test_framework/ripemd160.py:103]: nit: using the `&` operator is a slightly odd way to change the sign of an integer, no?

Might be better written as (119 - len&63) & 63 ?


See also https://docs.python.org/3/library/stdtypes.html#bitwise-operations-on-integer-types

> Performing these calculations with at least one extra sign extension bit in a finite two’s complement representation (a working bit-width of 1 + max(x.bit_length(), y.bit_length()) or more) is sufficient to get the same result as if there were an infinite number of sign bits.

**Review by maflcko — COMMENTED** [review]
Looks good and given that there are tests, this is probably correct. Left a nit to show that I read the code.

File: test/functional/test_framework/ripemd160.py (hunk 2 above)
Comment (sipa) [reply test/functional/test_framework/ripemd160.py:103]: I'm happy to change it, but "& 63" is exactly the same as "% 64" but faster, and that should be sufficient here.

Comment (sipa) [reply test/functional/test_framework/ripemd160.py:103]: Yes, the result of `&` is only negative if both inputs are negative.

File: test/functional/test_framework/ripemd160.py (hunk 1 above)
Comment (mmgen) [reply test/functional/test_framework/ripemd160.py:4]: > It's not constant time.

Given that its only use in Bitcoin is `ripemd160(sha256(data))`, does this not make a timing attack effectively useless?

Comment (sipa) [reply test/functional/test_framework/ripemd160.py:4]: @mmgen Yes, but it's open-source code, and open-source code gets copied and can end up being adopted for other purposes.

**Comment by richardkiss** [comment]
Thanks for this! I've ported to python 2 and integrated into pycoin here https://github.com/richardkiss/pycoin/commit/6b3e4bb3eeba4c41c33816533f7c52bbae7a6142

**Comment by fanquake** [comment]
Backported to 0.21 in #25538. 22.0 in #25250.

**Pull request merged**